cmd/podman/ps.go

```diff
@@ -71,12 +72,72 @@ type psJSONParams struct {
 	Namespaces       *batchcontainer.Namespace `json:"namespace,omitempty"`
 }
 
-// Type declaration and functions for sorting the PS output by time
+// Type declaration and functions for sorting the PS output
 type psSorted []psTemplateParams
 
-func (a psSorted) Len() int           { return len(a) }
-func (a psSorted) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
-func (a psSorted) Less(i, j int) bool { return a[i].CreatedAtTime.After(a[j].CreatedAtTime) }
+func (a psSorted) Len() int      { return len(a) }
+func (a psSorted) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+type psSortedCommand struct{ psSorted }
+
+func (a psSortedCommand) Less(i, j int) bool { return a.psSorted[i].Command < a.psSorted[j].Command }
+
+type psSortedCreated struct{ psSorted }
+
+func (a psSortedCreated) Less(i, j int) bool {
+	return a.psSorted[i].CreatedAtTime.After(a.psSorted[j].CreatedAtTime)
+}
+
+type psSortedId struct{ psSorted }
+
+func (a psSortedId) Less(i, j int) bool { return a.psSorted[i].ID < a.psSorted[j].ID }
+
+type psSortedImage struct{ psSorted }
+
+func (a psSortedImage) Less(i, j int) bool { return a.psSorted[i].Image < a.psSorted[j].Image }
+
+type psSortedLabels struct{ psSorted }
+
+func (a psSortedLabels) Less(i, j int) bool { return a.psSorted[i].Labels < a.psSorted[j].Labels }
+
+type psSortedMounts struct{ psSorted }
+
+func (a psSortedMounts) Less(i, j int) bool { return a.psSorted[i].Mounts < a.psSorted[j].Mounts }
+
+type psSortedNames struct{ psSorted }
+
+func (a psSortedNames) Less(i, j int) bool { return a.psSorted[i].Names < a.psSorted[j].Names }
+
+type psSortedPorts struct{ psSorted }
+
+func (a psSortedPorts) Less(i, j int) bool { return a.psSorted[i].Ports < a.psSorted[j].Ports }
+
+type psSortedRunningFor struct{ psSorted }
+
+func (a psSortedRunningFor) Less(i, j int) bool {
+	return a.psSorted[i].RunningFor < a.psSorted[j].RunningFor
+}
+
+type psSortedStatus struct{ psSorted }
+
+func (a psSortedStatus) Less(i, j int) bool { return a.psSorted[i].Status < a.psSorted[j].Status }
+
+type psSortedSize struct{ psSorted }
+
+func (a psSortedSize) Less(i, j int) bool {
+	r := regexp.MustCompile(`^\S+\s+\(virtual (\S+)\)`)
```

Comment:
Why are we doing this on the raw strings? We're getting raw uint64 sizes back, we should try and work with those instead of using regular expressions and string parsing on the strings we make with the ints.

Reply:
because the psTemplateParams only has size as a string, that was formatted on line 549. I figured it would be better to work with the data given than add extra data to the struct. I can definitely refactor if you think it would be better, though.

Reply:
@haircommander If we're going to be sorting off things that aren't in the `psTemplateParams`, we should probably sort `BatchContainerStruct` instead of `psTemplateParams`.

Reply:
Wouldn't that have more memory and time overhead? right now, each BatchContainerOp struct is a temp variable in a for loop on the containers (514), which are converted to psTemplateParams (which need to be returned to format the output). If we save the whole slice of BatchContainerOp, then sort it, then convert it to psTemplateParams, that's a whole extra slice of memory, plus a whole extra run through adding time. Here are my thoughts:

If we are prioritizing time, we should save the size in uint64 in the psTemplateParams, and use that to sort, instead of the regex stuff.
If we prioritize memory, do what is happening right now.

The whole refactor to BatchContainerStruct to me seems excessive in both benchmarks and also very time consuming. 

What do you think?

Reply:
@haircommander The difficulty with adding the int to psTemplateParams is that psTemplateParams is the literal output format for textual output. I think adding fields that are not directly displayed could interfere with the output code.

My other issue with sorting this late is that we won't be sorting the JSON output. I can see cases where we might want to do that.

On your output performance concern: I'll note that fully 75% of the time `ps` takes to execute right now is parsing and outputting the `psTemplateParams` struct...

Honestly this code needs to be rewritten extensively at some point. There are so many things in the batch struct that shouldn't be there that we waste time and memory copying out of the library.